File: Baekjoon/Implementation/[2910]빈도 정렬/solution.cpp
#include <algorithm>
#include <iostream>
#include <map>
#include <vector>

using namespace std;

typedef pair<int, int> pii;

#define endl "\n"

map<int, int> freq;
map<int, int> order;
vector<pii> sorted;
int N, C;

int cmp(pii a, pii b);

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  cin >> N >> C;

  for (int i = 0; i < N; i++) {
    int tmp;

    cin >> tmp;

    freq[tmp]++;

    if (order.find(tmp) == order.end()) {
      order[tmp] = i;
    }
  }

  for (pii p : freq) {
    sorted.push_back(p);
  }

  sort(sorted.begin(), sorted.end(), cmp);

  for (int i = 0; i < sorted.size(); i++) {
    for (int j = 0; j < sorted[i].second; j++) {
      cout << sorted[i].first << " ";
    }
  }
  cout << endl;

  return 0;
}

int cmp(pii a, pii b) {
  return a.second != b.second ? b.second < a.second : order[a.first] < order[b.first];
}

File: Baekjoon/Implementation/[2910]빈도 정렬/solution.py
import sys

N, C = map(int, sys.stdin.readline().split())
inputData = list(map(int, sys.stdin.readline().split()))

freq = dict()
order = dict()
sorted = list()

for i in range(N):
    if inputData[i] in freq:
        freq[inputData[i]] += 1
    else:
        freq[inputData[i]] = 1

    if inputData[i] not in order:
        order[inputData[i]] = i

for key, value in freq.items():
    sorted.append([key, value])

sorted.sort(key=lambda x: (-x[1], order[x[0]]))

for i in range(len(sorted)):
    for j in range(sorted[i][1]):
        sys.stdout.write(str(sorted[i][0]) + " ")
sys.stdout.write("\n")
